homeassistant/components/http/__init__.py

```diff
@@ -87,18 +90,88 @@
 STORAGE_VERSION: Final = 1
 SAVE_DELAY: Final = 180
 
+CONF_SERVERS = "servers"
+SERVERS_GROUP = "servers"
+
+SERVER_SCHEMA_WITHOUT_PORT = {
+    vol.Optional(CONF_SERVER_HOST): vol.All(
+        cv.ensure_list, vol.Length(min=1), [cv.string]
+    ),
+    vol.Optional(CONF_SSL_CERTIFICATE): cv.isfile,
+    vol.Optional(CONF_SSL_PEER_CERTIFICATE): cv.isfile,
+    vol.Optional(CONF_SSL_KEY): cv.isfile,
+    vol.Optional(CONF_SSL_PROFILE, default=SSL_MODERN): vol.In(
+        [SSL_INTERMEDIATE, SSL_MODERN]
+    ),
+}
+
+OPTIONAL_PORT = {vol.Optional(CONF_SERVER_PORT, default=SERVER_PORT): cv.port}
+
+_EXCLUSIVE_PORT_KEY = vol.Exclusive(CONF_SERVER_PORT, SERVERS_GROUP)
+_EXCLUSIVE_PORT_KEY.default = vol.default_factory(SERVER_PORT)
+
+EXCLUSIVE_PORT = {_EXCLUSIVE_PORT_KEY: cv.port}
+
+
+def _has_all_unique_ports(servers: list[dict[str, Any]]) -> list[dict[str, Any]]:
+    """Validate that each http service has a unique port."""
+    ports = [list[CONF_SERVER_PORT] for list in servers]
+    vol.Schema(vol.Unique())(ports)
+    return servers
+
+
+SERVERS_EXCLUSIVE_MESSAGE = (
+    'Configure one server at top level or configure multiple servers under "servers"'
+)
+
+
+def _relocated_with_message(key: str, new_location: str) -> Callable[[dict], dict]:
+    """Log key as relocated with a message."""
+
+    def validator(config: dict) -> dict:
+        """Check if key is in config and log the new location."""
+        near = ""
+        with contextlib.suppress(AttributeError):
+            near = (
+                f"near {config.__config_file__}"  # type: ignore[attr-defined]
+                f":{config.__line__}"
+            )
+        if key in config:
+            _LOGGER.warning(
+                "The '%s' option %s has moved to '%s', please update your configuration.",
+                key,
+                near,
+                new_location,
+            )
+        return config
+
+    return validator
+
+
 HTTP_SCHEMA: Final = vol.All(
     cv.deprecated(CONF_BASE_URL),
+    _relocated_with_message(CONF_SERVER_HOST, new_location="servers[0].server_host"),
```

Comment:
I originally marked these as deprecated, but I felt we needed to give better instructions as since I wanted to avoid a flurry of people getting the wrong impression and thinking ssl support was being removed.